cpu/stm32_common/periph/uart.c

```diff
@@ -139,6 +141,47 @@ int uart_init(uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
     return UART_OK;
 }
 
+int uart_mode(uart_t uart, uart_databits_t databits, uart_parity_t parity, uart_stopbits_t stopbits)
+{
+    assert(uart < UART_NUMOF);
+
+    uint32_t cr1;
+    uint32_t cr2;
+
+    if ((databits == 0x02)
+            || (databits == UART_DATABITS_7 && !parity)
+            || (databits == UART_DATABITS_9 && parity))
+        return UART_NOMODE;
+
+    if (parity == 0x02)
+        return UART_NOMODE;
+
+    cr1 = dev(uart)->CR1;
+    cr2 = dev(uart)->CR2;
+
+    if (parity) {
+	/* enable parity */
+        cr1 |= USART_CR1_PCE;
+        /* enable 9-bit mode for 8 databits or
+           mask the MSB for 7 databits mode */
+        if (databits == UART_DATABITS_8)
+            cr1 |= USART_CR1_M; 
```

Comment:
It seems there are some differences between the uart registers similar to the i2c registers the stm32l1xxx and stm32l0xxx are good examples.
In the l1 version the M register cannot be modified while transmitting (maybe add some protection for that) and the l0 has both a M0 and M1 bit (allowing data of 7,8,9).

Also the L0 says 'This bit can only be written when the USART is disabled (UE=0)' so maybe just shut off to change the mode.

[L1 example](https://www.st.com/content/ccc/resource/technical/document/reference_manual/cc/f9/93/b2/f0/82/42/57/CD00240193.pdf/files/CD00240193.pdf/jcr:content/translations/en.CD00240193.pdf) page 740
[L0 example](https://www.st.com/content/ccc/resource/technical/document/reference_manual/2f/b9/c6/34/28/29/42/d2/DM00095744.pdf/files/DM00095744.pdf/jcr:content/translations/en.DM00095744.pdf) page 801

Reply:
Should you #ifdef them around or does everything work out?

Reply:
I think you have a handle on it, it is actually nice to use the real macro then the #if STM32F0 | STM32F1 ...